docs/source/design/credentials.rst

```diff
@@ -28,11 +33,67 @@ These locations are tried in the following order, and the first successful looku
+References to credential and authenticator types are strings that are mapped to classes by the following dict definitions:
+
+- ``datalad.downloaders.AUTHENTICATION_TYPES``
+- ``datalad.downloaders.CREDENTIAL_TYPES``
+
+Available providers can be loaded by ``Providers.from_config_files`` and ``Providers.get_provider(url)`` will match a given URL against them and return the appropriate `Provider` instance.
+A :class:`Provider` object will determine a downloader to use (derived from :class:`BaseDownloader`), based on the URL's protocol.
+
+Note, that the provider config files are not currently following datalad's general config approach.
+Instead they are special config files, read by :class:`configparser.ConfigParser` that are not compatible with `git-config` and hence the :class:`ConfigManager`.
+
+There are currently two ways of storing a provider and thus creating its config file: ``Providers.enter_new`` and ``Providers._store_new``.
+The former will only work interactively and provide the user with options to choose from, while the latter is non-interactive and can therefore only be used, when all properties of the provider config are known and passed to it.
+There's no way at the moment to store an existing :class:`Provider` object directly.
+
+Integration with git
+====================
+
+In addition, there's a special case for interfacing `git-credential`: A dedicated :class:`GitCredential` class is used to talk to git's ``git-credential`` command instead of the keyring wrapper.
+This class has identical fields to the :class:`UserPassword` class and thus can be used by the same authenticators.
+Since git's way to deal with credentials doesn't involve labels but only matching URLs, it is - in some sense - the equivalent of datalad's provider layer.
+However, providers don't talk to a backend, credentials do.
+Hence, a more seemless integration requires some changes in the design of datalad's credential system as a whole.
+
+In the opposite direction - making git aware of datalad's credentials, there's no special casing, though.
+Datalad comes with a `git-credential-datalad` executable.
+Whenever git is configured to use it by setting `credential.helper=datalad`, it will be able to query datalad's credential system for a provider matching the URL in question and retrieve the referenced by this provider credentials.
+This helper can also store a new provider+credentials when asked to do so by git.
+It can do this interactively, asking a user to confirm/change that config or - if `credential.helper='datalad --non-interactive'` - try to non-interactively store with its defaults.
+
+Authenticators
+==============
+
+Authenticators are used by downloaders to issue authenticated requests.
+They are not easily available to directly be applied to requests being made outside of the downloaders.
+
+TODO
+
+Downloaders
+===========
+
+TODO
```

Comment:
I am not confident that there is value in leaving these TODO markers. We could have 10k of them. Why here?

Reply:
Right. Wasn't meant to be committed. Wanted to enhance those aspects outside of this PR.